### ISSUE TITLE
#1 Postęp prac - stan na 13.10.2020

### ISSUE THREAD

**Issue by wojtek11530** [wojtek11530/song_lyric_classification#12]
# Postęp prac - stan na 13.10.2020

## Lista wykonanych rzeczy
- skrypty tworzące zbiór danych:
  - skrypt pobierający surowe dane o piosenkach i emocjach
  - skrypt usuwający niepotrzebne dane ze zbiorów danych, łączący zbiory danych (normalizacja, standaryzacja, ekstrakcja kategorii dyskretnych (happy, relaxed, sad, angry) dla danych, dla których podane są wartości liczbowe  *Valance* i *Arousal*)  
  -  skrypt ściągający teksty piosenek korzystający z API LyricsGenius oraz ściągający gatunek muzyczny ze strony `theaudiodb.com`; przy pomocy biblioteki `langdetect` odfiltrowane są piosenki, które nie są w języku angielskim
  - skrypt filtrujący dane:
    - usunięcie ręcznie wybranych rekordów z niepoprawnymi, "śmieciowymi" tekstami piosenek
    - usunięcie rekordów z za długimi tekstami piosenek
    - usunięcie duplikatów piosenek
    - zgeneralizowanie gatunków muzycznych do 7 kategorii
  - skrypt dzielący zbiór danych na zbiór treningowy, walidacyjny oraz testowy
 - implementacja funkcji działających na tekście (wykorzystana biblioteka `NLTK`):
    - funkcja preprocesująca tekst:
       - usunięcie tekstów w nawiasach kwadratowych, np. [INTRO], nie będących faktycznym tekstem piosenki
       - zmiana liter na małe
       - usunięcie znaków interpunkcyjnych (oprócz apostrofu)
       - usunięcie cyfr
       - redukcja white space do pojedynczej spacji
    - funkcja usuwająca *stop words* (z wykluczeniem słów zawierających negacje)
    - funkcja wykonująca lematyzację tekstu
 - stworzenie modelu embeddingu Fasttext:
   - skrypt ściągający duży zbiór danych z `kaggle` zawierający 164790 tekstów piosenek: [*Song lyrics from 6 musical genres*](https://www.kaggle.com/neisse/scrapped-lyrics-from-6-genres?select=lyrics-data.csv)
   - skrypt tworzący model word embeddingu przy pomocy biblioteki `fasttext`
   - obecny model:
     - wykorzystane dane ze zbioru *Song lyrics from 6 musical genres*
     - zastosowany preprocessing tekstu i usunięciu *stop words*
     - wymiar embeddingu: 200
 - implementacja modeli klasyfikujących do 4 kategorii emocji (wykorzystane biblioteki: PyTorch, PyTorch Lightning):
   - model MLP:
     - jako wejście średnia ze wszystkich embeddingów słów w tekście
   - sieci rekurencyjne:
     - jako wejście sekwencja embeddingów dla `_max_num_words` pierwszych słów w tekście, jeśli sekwencja krótsza dodany paddding
     - model LSTM
     - model GRU
   - sieć konwolucyjna:
     -  jako wejście macierz embeddingów dla `_max_num_words` pierwszych słów w tekście, jeśli sekwencja krótsza dodany paddding
     - możliwość zastosowania filtrów różnej długości 
   - dostrojenie hiperarametrów ww. modeli w celu uzyskania najlepszych wyników
 - implementacja wariantów ww. modeli działających na poszczególnych fragmentach testu:
    - utworzenie funkcji dzielącej tekst piosenki na fragmenty, podział następuje po tekstach w nawiasach kwadratowych (np. [INTRO] lub [CHORUS]), tak żeby brać pod uwagę poszczególne zwrotki, refreny
   - jeśli w tekście brak tekstów w nawiasach kwadratowych, podział na fragmenty o stałej liczbie słów
   - w każdym z modeli procesowanie każdego fragmentu osobno, jako wyjście dla danej piosenki średnia wektorów wyjściowych dla poszczególnych fragmentów
   - nie wykonano jeszcze strojenia tego wariantu modeli, jednak jak dotąd wyniki gorsze niż dla standardowych modeli

## Wyniki

Dla każdego modelu teksty piosenek były poddane preprocesingowi oraz usunięto `stop words` wykorzystując zaimplementowane funkcje. Do ekstrakcji cech posłużył model Fast Text (opisany wyżej) wykonujący word embedding dla każdego słowa (tokena). 

Modele trenowane na zbiorze treningowym, trening przerwany, gdy maksymalna liczba epok osiągnięta lub gdy wartość funkcji straty na zbiorze walidacyjnym nie malała przez zadaną liczbę epok.

Dla każdego modelu parametry zostały dostrojone, przedstawione wyniki są uzyskane na zbiorze testowym.

### Model MLP

Warstwa wyjściowa: 200, warstwa ukryta: 24, warstwa wyjściowa: 4
dropout: 0.5
learning rate: 1e-3
weight decay: 1e-5
batch size: 128
Czas uczenia: __2m 11s__

```
              precision    recall  f1-score   support

       angry       0.66      0.62      0.64        56
       happy       0.62      0.84      0.71        94
     relaxed       0.50      0.37      0.43        51
         sad       0.45      0.32      0.37        60

    accuracy                           0.58       261
   macro avg       0.56      0.54      0.54       261
weighted avg       0.57      0.58      0.56       261
```
Confusion matrix:
![MLP_conf_matrix](https://user-images.githubusercontent.com/33195859/95856909-14f94e80-0d5b-11eb-9743-75330d30d440.png)

### Model LSTM

Warstwa wyjściowa: 200, warstwa ukryta: 200, warstwa wyjściowa: 4
Jedna warstwa ukryta, jednokierunkowa
dropout: 0.0
learning rate: 9e-5
weight decay: 1e-4
`max_num_words`: 200
batch size: 128
Czas uczenia: __10m 37s__

```
              precision    recall  f1-score   support

       angry       0.66      0.70      0.68        56
       happy       0.69      0.74      0.72        94
     relaxed       0.48      0.53      0.50        51
         sad       0.40      0.30      0.34        60

    accuracy                           0.59       261
   macro avg       0.56      0.57      0.56       261
weighted avg       0.58      0.59      0.58       261
```
Confusion matrix:
![LSTM_conf_matrix](https://user-images.githubusercontent.com/33195859/95857215-90f39680-0d5b-11eb-8cd9-c2594920fb63.png)

### Model GRU

Warstwa wyjściowa: 200, warstwa ukryta: 200, warstwa wyjściowa: 4
Jedna warstwa ukryta, jednokierunkowa
dropout: 0.0
learning rate: 9e-5
weight decay: 1e-5
`max_num_words`: 200
batch size: 128
Czas uczenia: __8m 09s__

```
              precision    recall  f1-score   support

       angry       0.63      0.73      0.68        56
       happy       0.67      0.83      0.74        94
     relaxed       0.53      0.39      0.45        51
         sad       0.41      0.28      0.34        60

    accuracy                           0.60       261
   macro avg       0.56      0.56      0.55       261
weighted avg       0.57      0.60      0.58       261
```
Confusion matrix:
![GRU_conf_matrix](https://user-images.githubusercontent.com/33195859/95857350-c6987f80-0d5b-11eb-89cd-da86117128bc.png)

#### Wersja 1

Warstwa wyjściowa: 200, warstwa wyjściowa: 4
Rozmiar filtrów: 5, 10 ,15
Liczba filtrów dla każdego rozmiaru: 64
dropout: 0.5
learning rate: 1e-4
weight decay: 0.0065
`max_num_words`: 200
batch size: 128
Czas uczenia: __9m 05s__

```
              precision    recall  f1-score   support

       angry       0.64      0.61      0.62        56
       happy       0.60      0.80      0.68        94
     relaxed       0.62      0.47      0.53        51
         sad       0.59      0.43      0.50        60

    accuracy                           0.61       261
   macro avg       0.61      0.58      0.59       261
weighted avg       0.61      0.61      0.60       261
```
Confusion matrix:
![ConvNet_conf_matrix](https://user-images.githubusercontent.com/33195859/95857741-5a6a4b80-0d5c-11eb-8dfb-cc3978f9777f.png)

#### Wersja 2

Warstwa wyjściowa: 200, warstwa wyjściowa: 4
Rozmiar filtrów: 5, 10 ,15
Liczba filtrów dla każdego rozmiaru: 128
dropout: 0.4
learning rate: 1e-4
weight decay: 0.0003
`max_num_words`: 200
batch size: 128
Czas uczenia: __6m 22s__

```
              precision    recall  f1-score   support

       angry       0.67      0.62      0.65        56
       happy       0.61      0.80      0.69        94
     relaxed       0.61      0.49      0.54        51
         sad       0.57      0.43      0.49        60

    accuracy                           0.62       261
   macro avg       0.62      0.59      0.59       261
weighted avg       0.61      0.62      0.61       261
```
Confusion matrix:
![ConvNet_v2_conf_matrix](https://user-images.githubusercontent.com/33195859/95857819-7a9a0a80-0d5c-11eb-8ccb-8ea8ea8b156c.png)

## Podsumowanie

Najlepsze wyniki otrzymane są dla sieci konwolucyjnej, __ok. 62%__ accuracy, przy niezłym f-scorze dla każdej kategorii. Sam model lepiej sobie radzi z piosenkami z kategorii o silnych emocjach (happy, angry) niż tych spokojniejszych (sad, relaxed).

Wynik dla sieci konwolucyjnej jest jednak tylko 4 punkty procentowe wyższy od prostego modelu MLP. Dla sieci rekurencyjnych przyrost skuteczności jest jeszcze mniejszy, a skutkuje znacznym wydłużeniem czasu trenowania.

Modele w wariantach opartych o fragmenty tekstu piosenki nie były jeszcze szczegółowiej dostrajane, jednak wstępne badania dają gorsze rezultaty.
